lib/r10k/tarball.rb

```diff
@@ -0,0 +1,257 @@
+        return false unless file_digest(found) == reader_digest(entry)
+      end
+      purge ? target_dir_entries.empty? : true
+    end
+
+    # Download the tarball from @source to @cache_path
+    def download
+      Dir.mktmpdir do
+        tempfile = Tempfile.new(tarball_basename)
+        tempfile.binmode
+        src_uri = URI.parse(source)
+
+        temp_digest = case src_uri.scheme
+                      when 'file', nil
+                        copy_to_file(src_uri.path, tempfile)
+                      when %r{[a-z]} # Windows drive letter
+                        copy_to_file(src_uri.to_s, tempfile)
+                      when %r{https?}
+                        download_to_file(src_uri, tempfile)
+                      else
+                        raise "Unexpected source scheme #{src_uri.scheme}"
+                      end
+
+        # Verify the download
+        unless (sha256digest == temp_digest) || sha256digest.nil?
+          raise 'Downloaded file does not match digest'
+        end
+
+        # Move the download to cache_path
+        FileUtils::mkdir_p(cache_root)
+        begin
+          FileUtils.mv(tempfile.path, cache_path)
+        rescue Errno::EACCES
+          FileUtils.cp(tempfile.path, cache_path)
+        end
+
+        tempfile.close
+        tempfile.unlink
+      end
+    end
+
+    # Is the cached tarball valid? Checks the cached tarball's digest against
+    # the expected digest.
+    #
+    # @return [Boolean]
+    def cache_valid?
+      unless sha256digest.nil?
+        File.exist?(cache_path) && sha256digest == file_digest(cache_path)
+      else
+        File.exist?(cache_path)
+      end
+    end
+
+    # List all of the files contained in the tarball and their paths. This is
+    # useful for implementing R10K::Purgable
+    #
+    # @return [Array] A list of file paths contained in the archive
+    def paths
+      names = Array.new
+      each_tarball_entry { |entry| names << entry.full_name.chomp('/') }
+      names
+    end
+
+    def cache_sha256digest
+      raise "Cache not present" unless File.exist?(cache_path)
+      file_digest(cache_path)
+    end
+
+    private
+
+    CHUNK_SIZE = 64 * 1024 # 64 kb
+
+    def each_tarball_entry(&block)
+      File.open(cache_path, 'rb') do |file|
+        Zlib::GzipReader.wrap(file) do |reader|
+          Archive::Tar::Minitar::Input.each_entry(reader) do |entry|
+            yield entry
+          end
+        end
+      end
+    end
+
+    # Return the sha256 digest of the file at the given path
+    #
+    # @param path [String] The path to the file
+    # @return [String] The file's sha256 hex digest
+    def file_digest(path)
+      File.open(path) do |file|
+        reader_digest(file)
+      end
+    end
+
+    # Return the sha256 digest of the readable data
+    #
+    # @param reader [String] An object that responds to #read
+    # @return [String] The read data's sha256 hex digest
+    def reader_digest(reader)
```

Comment:
Does it make sense to put these digest methods into some utility class that can be shared? They don't necessarily seem specific to this tarball feature.

Also I think they should include `sha256` in the names, since `digest` is fairly general.

Reply:
The names were originally left generic in order to be applicable in the event the digest algorithm ever changes, or is made selectable per-tarball. I've tried making that more evident by using a class constant and this comment. 

```
# At this time, the only checksum type supported is sha256. In the future,
# we may decide to support other algorithms if a use case arises. TBD.
DIGEST_CLASS = Digest::SHA256
```

The relevant methods all then look like

```
def reader_digest(reader)
  digest = DIGEST_CLASS.new
```

What do you think?